src/server/project.ts

```diff
@@ -1773,6 +1774,39 @@ namespace ts.server {
         getIncompleteCompletionsCache() {
             return this.projectService.getIncompleteCompletionsCache();
         }
+
+        /*@internal*/
+        withNoDtsResolutionProject(rootFileNames: string[], cb: (project: Project, ensureRoot: (fileName: string) => boolean) => void) {
+            const options: CompilerOptions = {
+                ...this.getCompilerOptions(),
+                noDtsResolution: true,
+                allowJs: true,
+                maxNodeModuleJsDepth: 3,
+                diagnostics: false,
+                skipLibCheck: true,
+                sourceMap: false,
+                types: ts.emptyArray,
+                lib: ts.emptyArray,
+                noLib: true,
+            };
+            const project = new SingleCommandProject(this.projectService, this.documentRegistry, options);
+            for (const fileName of rootFileNames) {
+                project.addRoot(this.getScriptInfo(fileName)!);
+            }
+            cb(
+                project,
+                fileName => {

```

Comment:
Why is this provided as a callback?  It looks like something the caller could handle since, AFAICT, it only closes over `project`?

Reply:
To make extra sure nobody tries to keep `project` open (or forgets to close it) beyond the synchronous execution of the single command.